Migrate from jsonwebtoken to jose to support Vercel environment

### Description of the Issue

I'm seeing the following error when trying to authenticate with JWT:

```
Cannot find module 'jsonwebtoken'
```

This happens when trying to load `jsonwebtoken` [here](https://github.com/box/box-typescript-sdk-gen/blob/2dd07d4d37913f7391cb594e0561636e73574a8d/src/utils.ts#L291). I tried adding installing `jsonwebtoken` to my `package.json` file, but the error remained.

My application is a Next.js app using the app router, uses `bun install` to install dependencies, and has `"type": "module"` in `package.json`. It seems like this could be a CJS vs ESM issue.

This issue has existed since v0.1.8 of `box-typescript-sdk-gen`.

### Versions Used

Typescript SDK: v0.3.0
Platform: Node.js
Node.js: 20.11.0

@michaelhays, the reason you are getting `invalid_resource` error when trying to downscope the token is a small bug in you authentication code. Instead of:

```typescript
    const boxJwtConfig = JwtConfig.fromConfigJsonString(process.env.BOX_CONFIG);
    const boxJwtAuth = new BoxJwtAuth({ config: boxJwtConfig });
    boxJwtAuth.asUser(process.env.BOX_APP_USER_ID);
``` 

you should do:

```typescript
    const boxJwtConfig = JwtConfig.fromConfigJsonString(process.env.BOX_CONFIG);
    const boxJwtAuth = await new BoxJwtAuth({ config: boxJwtConfig }).asUser(
      process.env.BOX_APP_USER_ID
    );
``` 

Please notice that `asUser` method does not mutate the instance of `BoxJwtAuth` - it creates a new one with updated settings. This function is currently asynchronous and will be made synchronous in the next release - you can `await` it anyway.

Using the updated code in your project, I was able to get the token correctly using both the legacy and new SDK. Thanks for providing the sample code.

As per the original issue "Cannot find module 'jsonwebtoken'", I wasn't able to reproduce it. If come across a sample project&env settings that cause this problem, please reopen this issue and we'll be happy to dig into it.

Hey @mhagmajer, thanks so much! Right you are, I made that change and it's working as expected locally :)

I deployed to Vercel and am seeing the `Cannot find module 'jsonwebtoken'` error now.

Vercel is the only environment I'm seeing the error in; it's working as expected every way I've tested locally:

1. `bun dev`
2. `bun run build` + `bun start`
3. `bun vercel dev`

A couple of things I've tried:
1. ([Commit](https://github.com/michaelhays/box-typescript-jsonwebtoken/commit/ef245e89782b38217a2004f1216c75a796703faf)) Directly adding the jsonwebtoken dependency
2. ([Commit](https://github.com/michaelhays/box-typescript-jsonwebtoken/commit/bd4863387834fc7dafb3be9246e3f53e60911973)) Importing the new library with `createRequire` to force CJS

If you'd like, I'm happy to help you get this deployed on Vercel yourself -- it's quite easy, only takes a couple of minutes.

(I also don't think I'm able to reopen this issue myself, so you may have to do it on your end)

--- a/.codegen.json
+++ b/.codegen.json
@@ -1,1 +1,1 @@
-{ "engineHash": "a8b56d7", "specHash": "b2f7568", "version": "0.5.0" }
+{ "engineHash": "c714d1b", "specHash": "b2f7568", "version": "0.5.0" }

--- a/package.json
+++ b/package.json
@@ -29,7 +29,7 @@
     "form-data": "^4.0.0",
     "node-fetch": "^2.6.3",
     "buffer": "^6.0.3",
-    "jsonwebtoken": "^9.0.0",
+    "jose": "^5.2.2",
     "tslib": "^2.6.2",
     "uuid": "^9.0.0"
   },
@@ -38,7 +38,6 @@
     "@rollup/plugin-node-resolve": "^15.2.3",
     "@rollup/plugin-typescript": "^11.1.5",
     "@types/jest": "^29.5.1",
-    "@types/jsonwebtoken": "^9.0.1",
     "@types/node": "*",
     "@types/node-fetch": "2.6.3",
     "@types/uuid": "^9.0.1",

--- a/src/box/jwtAuth.generated.ts
+++ b/src/box/jwtAuth.generated.ts
@@ -159,7 +159,11 @@ export class BoxJwtAuth implements Authentication {
       key: this.config.privateKey,
       passphrase: this.config.privateKeyPassphrase,
     } satisfies JwtKey;
-    const assertion: string = createJwtAssertion(claims, jwtKey, jwtOptions);
+    const assertion: string = await createJwtAssertion(
+      claims,
+      jwtKey,
+      jwtOptions
+    );
     const authManager: AuthorizationManager = !(networkSession == void 0)
       ? new AuthorizationManager({ networkSession: networkSession })
       : new AuthorizationManager({});

--- a/src/internal/utils.ts
+++ b/src/internal/utils.ts
@@ -1,6 +1,7 @@
 import { Buffer } from 'buffer';
 import type { Readable } from 'stream';
 import { v4 as uuidv4 } from 'uuid';
+import { SignJWT, importPKCS8 } from 'jose';
 
 export function isBrowser() {
   return (
@@ -266,11 +267,6 @@ export type JwtSignOptions = {
   subject?: string | undefined;
   issuer?: string | undefined;
   jwtid?: string | undefined;
-  mutatePayload?: boolean | undefined;
-  noTimestamp?: boolean | undefined;
-  encoding?: string | undefined;
-  allowInsecureKeySizes?: boolean | undefined;
-  allowInvalidAsymmetricKeyTypes?: boolean | undefined;
 };
 
 /**
@@ -281,15 +277,36 @@ export type JwtSignOptions = {
  * @param options
  * @returns
  */
-export function createJwtAssertion(
+export async function createJwtAssertion(
   claims: {
     readonly [key: string]: any;
   },
   key: JwtKey,
   options: JwtSignOptions
-): string {
-  const jwt = eval('require')('jsonwebtoken');
-  return jwt.sign(claims, key, options);
+): Promise<string> {
+  const crypto = eval('require')('crypto');
+  const privateKey = crypto.createPrivateKey({
+    key: key.key,
+    format: 'pem',
+    type: 'pkcs8',
+    passphrase: key.passphrase,
+  });
+  const pem = privateKey.export({ type: 'pkcs8', format: 'pem' }).toString();
+  const pkcs8 = await importPKCS8(pem, options.algorithm || 'RS256');
+  let signer = new SignJWT(claims);
+  signer = options.audience ? signer.setAudience(options.audience) : signer;
+  signer = options.expiresIn
+    ? signer.setExpirationTime(options.expiresIn)
+    : signer;
+  signer = options.issuer ? signer.setIssuer(options.issuer) : signer;
+  signer = options.jwtid ? signer.setJti(options.jwtid) : signer;
+  signer = options.notBefore ? signer.setNotBefore(options.notBefore) : signer;
+  signer = options.subject ? signer.setSubject(options.subject) : signer;
+  signer = options.algorithm
+    ? signer.setProtectedHeader({ alg: options.algorithm })
+    : signer;
+  signer = signer.setIssuedAt();
+  return await signer.sign(pkcs8);
 }
 
 /**

